### PR TITLE
Update wdio-browserstack-service for insights (v7)

### PR DESCRIPTION
**Pull request by nagpalkaran95**
## Proposed changes

BrowserStack wants to collect test related data through the BrowserStack service for WDIO to enhance the reporting capabilities to the users of BrowserStack + WebdriverIO.
The test data that would be collected through the service includes test name, test organisation hierarchy, file name & structure, status, error stacktrace, screenshots, test commands, source control info about the test code, a few basic host/CI level info, etc.

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

Related PRs:
  * v8 - https://github.com/webdriverio/webdriverio/pull/9430
  * v7 (closed due to CLA issue) - https://github.com/webdriverio/webdriverio/pull/9124

### CODE REVIEW

**Review by Ankit098 — APPROVED** [review]

**Comment by nagpalkaran95** [comment]
@christian-bromann Raised a new PR as discussed. Have linked the old one for reference. That can be closed now.

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+        }
+        await this.sendTestRunEvent(test, 'TestRunFinished', result)
+    }
+
+    /**
+      * Cucumber Only
+      */
+
+    async beforeScenario (world: ITestCaseHookParameter) {
+        let pickleData = world.pickle
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:95]: I personally would not create variables for these kind of things given there doesn't seem to be a need for it but I don't want to be too picky here. However if we create variables we don't modify we should at least define them as `const`
```suggestion
        const pickleData = world.pickle
```

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+            testMetaData = {
+                steps: []
+            }
+        }
+
+        if (testMetaData && !testMetaData['steps']) {
+            testMetaData['steps'] = []
+        }
+
+        testMetaData['steps']?.push({
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:142]:
```suggestion
        testMetaData.steps?.push({
```

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        const uniqueId = getUniqueIdentifierForCucumber({ pickle: scenario } as ITestCaseHookParameter)
+        let testMetaData = this._tests[uniqueId]
+        if (!testMetaData) {
+            testMetaData = {
+                steps: []
+            }
+        }
+
+        if (testMetaData && !testMetaData['steps']) {
+            testMetaData['steps'] = []
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:139]:
```suggestion
            testMetaData.steps = []
```

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+        waitTimeout = DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS,
+        waitInterval = DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS
+    ) {
+        if (this._requestQueueHandler.pendingUploads <= 0 || waitTimeout <= 0) {
+            return
+        }
+
+        await sleep(waitInterval)
+        this.uploadPending(waitTimeout - waitInterval)
+        return
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:193]:
```suggestion
        return this.uploadPending(waitTimeout - waitInterval)
```

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+
+        const { feature, scenario, steps, uuid, startedAt, finishedAt } = this._tests[uniqueId] || {}
+
+        let fullNameWithExamples = world.pickle.name
+        const examples = getScenarioExamples(world)
+        if (examples) {
+            fullNameWithExamples = world.pickle.name + ' (' + examples.join(', ')  + ')'
+        }
+
+        /* istanbul ignore next */
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:364]: I don't see why we should ignore the coverage of the next part

```suggestion
```

File: packages/wdio-browserstack-service/src/insights-handler.ts [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+
+        const req = this._requestQueueHandler.add(uploadData)
+
+        if (req.proceed && req.data) {
+            await uploadEventData(req.data, req.url)
+        }
+    }
+
+    private getIntegrationsObject () {
+        /* istanbul ignore next */
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/insights-handler.ts:440]: Same here, it seems like an easy unit test to write if it breaks coverage
```suggestion
```

File: packages/wdio-browserstack-service/src/reporter.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    private _sessionId?: string
+    private _suiteName?: string
+    private _requestQueueHandler = RequestQueueHandler.getInstance()
+
+    onRunnerStart (runnerStats: RunnerStats) {
+        this._capabilities = runnerStats.capabilities as Capabilities.Capabilities
+        this._config = runnerStats.config as BrowserstackConfig & Options.Testrunner
+        this._sessionId = runnerStats.sessionId
+        /* istanbul ignore next */
+        if (this._config.testObservability == false) this._observability = false
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/reporter.ts:24]: Never use `==` comparisons
```suggestion
        this._observability = this._config.testObservability
```

File: packages/wdio-browserstack-service/src/reporter.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    onSuiteStart (suiteStats: SuiteStats) {
+        this._suiteName = suiteStats.file
+    }
+
+    async onTestSkip (testStats: TestStats) {
+        // cucumber steps call this method. We don't want step skipped state so skip for cucumber
+
+        /* istanbul ignore next */
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/reporter.ts:34]:
```suggestion
```

File: packages/wdio-browserstack-service/src/reporter.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        const framework = this._config?.framework
+
+        if (this._observability && framework != 'cucumber') {
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/reporter.ts:37]:
```suggestion
        if (this._observability && framework !== 'cucumber') {
```

File: packages/wdio-browserstack-service/src/util.ts [hunk 10]
```diff
@@ -1,7 +1,37 @@
+import { hostname, platform, type, version, arch } from 'os'
+import { promisify } from 'util'
+import * as http from 'http'
+import * as https from 'https'
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/util.ts:4]:
```suggestion
import http from 'http'
import https from 'https'
```

File: packages/wdio-browserstack-service/src/util.ts [hunk 11]
```diff
@@ -1,7 +1,37 @@
...
+import path from 'path'
+
 import type { Browser, MultiRemoteBrowser } from 'webdriverio'
-import type { Capabilities } from '@wdio/types'
+import type { Capabilities, Frameworks, Options } from '@wdio/types'
+import { BeforeCommandArgs, AfterCommandArgs } from '@wdio/reporter'
+import logger from '@wdio/logger'
+
+import got from 'got'
+import { HTTPError } from 'got'
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/util.ts:13]:
```suggestion
import got, { HTTPError } from 'got'
```

**Review by christian-bromann — COMMENTED** [review]
Did another round of reviews. I will make sure to add more eslint rules to make some of the coding conventions more obvious.

**Comment by nagpalkaran95** [comment]
@christian-bromann getting unrelated failure for windows..possible to re-run the tests?

File: packages/wdio-browserstack-service/src/service.ts [hunk 12]
```diff
@@ -27,6 +35,12 @@ export default class BrowserstackService implements Services.ServiceInstance {
         this._options = { ...DEFAULT_OPTIONS, ...options }
         // added to maintain backward compatibility with webdriverIO v5
         this._config || (this._config = this._options)
+        this._observability = this._options.testObservability
+
+        if (this._observability) {
+            this._config.reporters ? this._config.reporters.push(path.join(__dirname, 'reporter.js')) : [path.join(__dirname, 'reporter.js')]
+        }
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/service.ts:42]:
```suggestion
        if (this._observability) {
            this._config.reporters.push(path.join(__dirname, 'reporter.js'))
        }
```

File: packages/wdio-browserstack-service/src/service.ts [hunk 13]
```diff
@@ -179,6 +221,18 @@ export default class BrowserstackService implements Services.ServiceInstance {
...
             this._failReasons.push(exception)
         }
+
+        await this._insightsHandler?.afterScenario(world)
+    }
+
+    async beforeStep (step: Frameworks.PickleStep, scenario: Pickle) {
+        const { keyword, text } = step
+        await this._insightsHandler?.beforeStep(step, scenario)
+        await this._setAnnotation(`Step: ${keyword}${text}`)
```
Comment (christian-bromann) [inline packages/wdio-browserstack-service/src/service.ts:231]: There is not really any value pulling out properties out of an object if these are only used once. It only blows up the code base.
```suggestion
        await this._insightsHandler?.beforeStep(step, scenario)
        await this._setAnnotation(`Step: ${step.keyword}${step.text}`)
```

**Review by christian-bromann — COMMENTED** [review]
I think these will be the last issues I found.

**Comment by nagpalkaran95** [comment]
@christian-bromann added the same changes to the other PR (against main) as well.

**Review by christian-bromann — APPROVED** [review]
I will go ahead and merge this one. Thanks for pulling through the review cycles. As a result we have improved our EsLint settings in the `main` branch. Given this was a larger PR I wanted to make sure we adhere to the coding standards of the project. If you have feedback to our review process, please don't hesitate to send me an email.

LGTM 👍

**Pull request merged**